docs/man/memcached_behavior.3
.TH "MEMCACHED_BEHAVIOR" "3" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memcached_behavior \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
Manipulate the behavior of a memcached_st structure.
.SH SYNOPSIS
.sp
#include <libmemcached/memcached.h>
.INDENT 0.0
.TP
.B memcached_behavior_t
.UNINDENT
.INDENT 0.0
.TP
.B uint64_t memcached_behavior_get(memcached_st\fI\ *ptr\fP, \fI\%memcached_behavior_t\fP\fI\ flag\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_behavior_set(memcached_st\fI\ *ptr\fP, \fI\%memcached_behavior_t\fP\fI\ flag\fP, uint64_t\fI\ data\fP)
.UNINDENT
.sp
Compile and link with \-lmemcached
.SH DESCRIPTION
.sp
\fIlibmemcached(3)\fP behavior can be modified by using \fI\%memcached_behavior_set()\fP. Default behavior is the library strives to be quick and
accurate. Some behavior, while being faster, can also result in not entirely
accurate behavior (for instance, \fBmemcached_set()\fP will always respond
with \fBMEMCACHED_SUCCESS\fP).
.sp
\fI\%memcached_behavior_get()\fP takes a behavior flag and returns whether or not that behavior is currently enabled in the client.
.sp
\fI\%memcached_behavior_set()\fP changes the value of a particular option
of the client. It takes both a flag (listed below) and a value. For simple
on or off options you just need to pass in a value of 1. Calls to
\fI\%memcached_behavior_set()\fP will flush and reset all connections.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_USE_UDP
.UNINDENT
.sp
Causes \fIlibmemcached(3)\fP to use the UDP transport when communicating
with a memcached server. Not all I/O operations are testsed
when this behavior is enababled. The following operations will return
\fBMEMCACHED_NOT_SUPPORTED\fP when executed with the
\fI\%MEMCACHED_BEHAVIOR_USE_UDP\fP enabled: \fBmemcached_version()\fP,
\fBmemcached_stat()\fP, \fBmemcached_get()\fP,
\fBmemcached_get_by_key()\fP, \fBmemcached_mget()\fP,
\fBmemcached_mget_by_key()\fP, \fBmemcached_fetch()\fP,
\fBmemcached_fetch_result()\fP, \fBmemcached_fetch_execute()\fP.
.sp
All other operations are testsed but are executed in a \(aqfire\-and\-forget\(aq
mode, in which once the client has executed the operation, no attempt
will be made to ensure the operation has been received and acted on by the
server.
.sp
\fIlibmemcached(3)\fP does not allow TCP and UDP servers to be shared
within the same libmemached(3) client \(aqinstance\(aq. An attempt to add a TCP
server when this behavior is enabled will result in a \fBMEMCACHED_INVALID_HOST_PROTOCOL\fP, as will attempting to add a UDP server when this behavior has
not been enabled.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_NO_BLOCK
.UNINDENT
.sp
Causes \fIlibmemcached(3)\fP to use asychronous IO. This is the fastest
transport available for storage functions.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SND_TIMEOUT
.UNINDENT
.sp
This sets the microsecond behavior of the socket against the SO_SNDTIMEO flag.  In cases where you cannot use non\-blocking IO this will allow you to still have timeouts on the sending of data.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_RCV_TIMEOUT
.UNINDENT
.sp
This sets the microsecond behavior of the socket against the SO_RCVTIMEO flag.
.sp
In cases where you cannot use non\-blocking IO this will allow you to still have timeouts on the reading of data.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_TCP_NODELAY
.UNINDENT
.sp
Turns on the no\-delay feature for connecting sockets (may be faster in some
environments).
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_HASH
.UNINDENT
.sp
Makes the default hashing algorithm for keys use MD5. The value can be set to either \fBMEMCACHED_HASH_DEFAULT\fP, \fBMEMCACHED_HASH_MD5\fP, \fBMEMCACHED_HASH_CRC\fP, \fBMEMCACHED_HASH_FNV1_64\fP, \fBMEMCACHED_HASH_FNV1A_64\fP, \fBMEMCACHED_HASH_FNV1_32\fP, \fBMEMCACHED_HASH_FNV1A_32\fP, \fBMEMCACHED_HASH_JENKINS\fP, \fBMEMCACHED_HASH_HSIEH\fP, and \fBMEMCACHED_HASH_MURMUR\fP.
.sp
Each hash has it\(aqs advantages and it\(aqs weaknesses. If you don\(aqt know or don\(aqt
care, just go with the default.
.sp
Support for \fBMEMCACHED_HASH_HSIEH\fP is a compile time option that is disabled by default. To enable tests for this hashing algorithm, configure and build libmemcached with the \-\-enable\-hash_hsieh.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_DISTRIBUTION
.UNINDENT
.sp
Using this you can enable different means of distributing values to servers.
.sp
The default method is MEMCACHED_DISTRIBUTION_MODULA. You can enable consistent hashing by setting MEMCACHED_DISTRIBUTION_CONSISTENT.  Consistent hashing delivers better distribution and allows servers to be added to the cluster with minimal cache losses. Currently MEMCACHED_DISTRIBUTION_CONSISTENT is an alias for the value MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_CACHE_LOOKUPS
.UNINDENT
.sp
Deprecated since version 0.46(?): DNS lookups are now always cached until an error occurs with the server.
.sp
Memcached can cache named lookups so that DNS lookups are made only once.

.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SUPPORT_CAS
.UNINDENT
.sp
Support CAS operations (this is not enabled by default at this point in the server since it imposes a slight performance penalty).
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_KETAMA
.UNINDENT
.sp
Sets the default distribution to MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA and
the hash to \fBMEMCACHED_HASH_MD5\fP.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_KETAMA_WEIGHTED
Sets the default distribution to MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA with the weighted tests.  and the hash to MEMCACHED_HASH_MD5.
.UNINDENT
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_KETAMA_HASH
.UNINDENT
.sp
Sets the hashing algorithm for host mapping on continuum. The value can be set
to either \fBMEMCACHED_HASH_DEFAULT\fP, \fBMEMCACHED_HASH_MD5\fP,
\fBMEMCACHED_HASH_CRC\fP, \fBMEMCACHED_HASH_FNV1_64\fP,
\fBMEMCACHED_HASH_FNV1A_64\fP, \fBMEMCACHED_HASH_FNV1_32\fP, and
\fBMEMCACHED_HASH_FNV1A_32\fP.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_KETAMA_COMPAT
.UNINDENT
.sp
Sets the compatibility mode. The value can be set to either MEMCACHED_KETAMA_COMPAT_LIBMEMCACHED (this is the default) or MEMCACHED_KETAMA_COMPAT_SPY to be compatible with the SPY Memcached client for Java.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_POLL_TIMEOUT
.UNINDENT
.sp
Modify the timeout value that is used by poll. The default value is \-1. An signed int must be passed to memcached_behavior_set to change this value (this requires casting). For memcached_behavior_get a signed int value will be cast and returned as the unsigned long long.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_USER_DATA
.UNINDENT
.sp
Deprecated since version <: 0.30
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_BUFFER_REQUESTS
.UNINDENT
.sp
Enabling buffered IO causes commands to "buffer" instead of being sent. Any action that gets data causes this buffer to be be sent to the remote connection. Quiting the connection or closing down the connection will also cause the buffered data to be pushed to the remote connection.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_VERIFY_KEY
.UNINDENT
.sp
Enabling this will cause \fIlibmemcached(3)\fP to test all keys to verify that they are valid keys.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SORT_HOSTS
.UNINDENT
.sp
Enabling this will cause hosts that are added to be placed in the host list in sorted order. This will defeat consisten hashing.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_CONNECT_TIMEOUT
.UNINDENT
.sp
In non\-blocking mode this changes the value of the timeout during socket connection in milliseconds. Specifying \-1 means an infinite time‐out.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_BINARY_PROTOCOL
.UNINDENT
.sp
Enable the use of the binary protocol. Please note that you cannot toggle this flag on an open connection.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT
.UNINDENT
.sp
Set this value to enable the server be removed after continuous MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT times connection failure.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_IO_MSG_WATERMARK
.UNINDENT
.sp
Set this value to tune the number of messages that may be sent before libmemcached should start to automatically drain the input queue. Setting this value to high, may cause libmemcached to deadlock (trying to send data, but the send will block because the input buffer in the kernel is full).
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_IO_BYTES_WATERMARK
.UNINDENT
.sp
Set this value to tune the number of bytes that may be sent before libmemcached should start to automatically drain the input queue (need at least 10 IO requests sent without reading the input buffer). Setting this value to high, may cause libmemcached to deadlock (trying to send data, but the send will block because the input buffer in the kernel is full).
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_IO_KEY_PREFETCH
.UNINDENT
.sp
The binary protocol works a bit different than the textual protocol in that a multiget is implemented as a pipe of single get\-operations which are sent to the server in a chunk. If you are using large multigets from your application, you may improve the latency of the gets by setting this value so you send out the first chunk of requests when you hit the specified limit.  It allows the servers to start processing the requests to send the data back while the rest of the requests are created and sent to the server.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_NOREPLY
.UNINDENT
.sp
Set this value to specify that you really don\(aqt care about the result from your storage commands (set, add, replace, append, prepend).
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_NUMBER_OF_REPLICAS
.UNINDENT
.sp
If you just want "a poor mans HA", you may specify the numbers of replicas libmemcached should store of each item (on different servers).  This replication does not dedicate certain memcached servers to store the replicas in, but instead it will store the replicas together with all of the other objects (on the \(aqn\(aq next servers specified in your server list).
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_RANDOMIZE_REPLICA_READ
.UNINDENT
.sp
Allows randomizing the replica reads starting point. Normally the read is done from primary server and in case of miss the read is done from primary + 1, then primary + 2 all the way to \(aqn\(aq replicas. If this option is set on the starting point of the replica reads is randomized between the servers.  This allows distributing read load to multiple servers with the expense of more write traffic.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_CORK
.UNINDENT
.sp
This open has been deprecated with the behavior now built and used appropriately on selected platforms.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_KEEPALIVE
.UNINDENT
.sp
Enable TCP_KEEPALIVE behavior.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_KEEPALIVE_IDLE
.UNINDENT
.sp
Specify time, in seconds, to mark a connection as idle. This is only available as an option Linux.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SOCKET_SEND_SIZE
.UNINDENT
.sp
Find the current size of SO_SNDBUF. A value of 0 means either an error occured or no hosts were available. It is safe to assume system default if this occurs. If an error occurs you can checked the last cached errno statement to find the specific error.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SOCKET_RECV_SIZE
.UNINDENT
.sp
Find the current size of SO_RCVBUF. A value of 0 means either an error occured or no hosts were available. It is safe to assume system default if this occurs. If an error occurs you can checked the last cached errno statement to find the specific error.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT
.UNINDENT
.sp
Deprecated since version 0.48: See \fI\%MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS\fP
.sp
This number of times a host can have an error before it is disabled.

.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_AUTO_EJECT_HOSTS
.UNINDENT
.sp
Deprecated since version 0.48: See \fI\%MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS\fP
.sp
If enabled any hosts which have been flagged as disabled will be removed from the list of servers in the memcached_st structure. This must be used in combination with MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT.

.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS
If enabled any hosts which have been flagged as disabled will be removed from the list of servers in the memcached_st structure.
.UNINDENT
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_RETRY_TIMEOUT
.UNINDENT
.sp
When enabled a host which is problematic will only be checked for usage based on the amount of time set by this behavior. The value is in seconds.
.INDENT 0.0
.TP
.B MEMCACHED_BEHAVIOR_HASH_WITH_PREFIX_KEY
.UNINDENT
.sp
When enabled the prefix key will be added to the key when determining server
by hash. See \fBMEMCACHED_CALLBACK_NAMESPACE\fP for additional
information.
.SH RETURN
.sp
memcached_behavior_get returns either the current value of the get, or 0
or 1 on simple flag behaviors (1 being enabled). memcached_behavior_set
returns failure or success.
.SH NOTES
.sp
memcached_behavior_set in version .17 was changed from taking a pointer
to data value, to taking a uin64_t.
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP \fImemcached_strerror(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

libmemcached/memcached_behavior.c
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary: Change the behavior of the memcached connection.
 *
 */

#include "common.h"
#include <time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/tcp.h>

/*
  This function is used to modify the behavior of running client.

  We quit all connections so we can reset the sockets.
*/

memcached_return_t memcached_behavior_set(memcached_st *ptr,
                                          memcached_behavior_t flag,
                                          uint64_t data)
{
  switch (flag)
  {
  case MEMCACHED_BEHAVIOR_NUMBER_OF_REPLICAS:
    ptr->number_of_replicas= (uint32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_IO_MSG_WATERMARK:
    ptr->io_msg_watermark= (uint32_t) data;
    break;
  case MEMCACHED_BEHAVIOR_IO_BYTES_WATERMARK:
    ptr->io_bytes_watermark= (uint32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_IO_KEY_PREFETCH:
    ptr->io_key_prefetch = (uint32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_SND_TIMEOUT:
    ptr->snd_timeout= (int32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_RCV_TIMEOUT:
    ptr->rcv_timeout= (int32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT:
    ptr->server_failure_limit= (uint32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_BINARY_PROTOCOL:
    if (data)
        ptr->flags.verify_key= false;

    ptr->flags.binary_protocol= data ? true : false;
    break;
  case MEMCACHED_BEHAVIOR_SUPPORT_CAS:
    ptr->flags.support_cas= data ? true: false;
    break;
  case MEMCACHED_BEHAVIOR_NO_BLOCK:
    ptr->flags.no_block= data ? true: false;
    memcached_quit(ptr);
    break;
  case MEMCACHED_BEHAVIOR_BUFFER_REQUESTS:
    ptr->flags.buffer_requests= data ? true : false;
    memcached_quit(ptr);
    break;
  case MEMCACHED_BEHAVIOR_USE_UDP:
    if (ptr->number_of_hosts)
      return MEMCACHED_FAILURE;
    ptr->flags.use_udp= data ? true : false;

    if (data)
      ptr->flags.no_reply= data ? true : false;
    break;

  case MEMCACHED_BEHAVIOR_TCP_NODELAY:
    ptr->flags.tcp_nodelay= data ? true : false;
    memcached_quit(ptr);
    break;
  case MEMCACHED_BEHAVIOR_DISTRIBUTION:
    {
      ptr->distribution= (memcached_server_distribution_t)(data);
      if (ptr->distribution == MEMCACHED_DISTRIBUTION_RANDOM)
      {
        srandom((uint32_t) time(NULL));
      }
      run_distribution(ptr);
      break;
    }
  case MEMCACHED_BEHAVIOR_KETAMA:
    {
      if (data)
      {
        ptr->hash= MEMCACHED_HASH_MD5;
        ptr->distribution= MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA;
      }
      else
      {
        ptr->hash= 0;
        ptr->distribution= 0;
      }
      run_distribution(ptr);
      break;
    }
  case MEMCACHED_BEHAVIOR_KETAMA_WEIGHTED:
    {
      ptr->hash= MEMCACHED_HASH_MD5;
      ptr->distribution= MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA;
      ptr->flags.ketama_weighted= data ? true : false;
      run_distribution(ptr);
      break;
    }
  case MEMCACHED_BEHAVIOR_KETAMA_COMPAT_MODE:
    switch (data)
    {
    case MEMCACHED_KETAMA_COMPAT_LIBMEMCACHED:
      ptr->hash= MEMCACHED_HASH_MD5;
      ptr->distribution= MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA;
      break;
    case MEMCACHED_KETAMA_COMPAT_SPY:
      ptr->hash= MEMCACHED_HASH_MD5;
      ptr->distribution= MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA_SPY;
      break;
    default:
       return MEMCACHED_FAILURE;
    }
    run_distribution(ptr);
    break;
  case MEMCACHED_BEHAVIOR_HASH:
#ifndef HAVE_HSIEH_HASH
    if ((memcached_hash_t)(data) == MEMCACHED_HASH_HSIEH)
      return MEMCACHED_FAILURE;
#endif
    ptr->hash= (memcached_hash_t)(data);
    break;
  case MEMCACHED_BEHAVIOR_KETAMA_HASH:
    ptr->hash_continuum= (memcached_hash_t)(data);
    run_distribution(ptr);
    break;
  case MEMCACHED_BEHAVIOR_CACHE_LOOKUPS:
    ptr->flags.use_cache_lookups= data ? true : false;
    memcached_quit(ptr);
    break;
  case MEMCACHED_BEHAVIOR_VERIFY_KEY:
    if (ptr->flags.binary_protocol)
        break;
    ptr->flags.verify_key= data ? true : false;
    break;
  case MEMCACHED_BEHAVIOR_SORT_HOSTS:
    {
      ptr->flags.use_sort_hosts= data ? true : false;
      run_distribution(ptr);

      break;
    }
  case MEMCACHED_BEHAVIOR_POLL_TIMEOUT:
    ptr->poll_timeout= (int32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_CONNECT_TIMEOUT:
    ptr->connect_timeout= (int32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_RETRY_TIMEOUT:
    ptr->retry_timeout= (int32_t)data;
    break;
  case MEMCACHED_BEHAVIOR_SOCKET_SEND_SIZE:
    ptr->send_size= (int32_t)data;
    memcached_quit(ptr);
    break;
  case MEMCACHED_BEHAVIOR_SOCKET_RECV_SIZE:
    ptr->recv_size= (int32_t)data;
    memcached_quit(ptr);
    break;
  case MEMCACHED_BEHAVIOR_USER_DATA:
    return MEMCACHED_FAILURE;
  case MEMCACHED_BEHAVIOR_HASH_WITH_PREFIX_KEY:
    ptr->flags.hash_with_prefix_key= data ? true : false;
    break;
  case MEMCACHED_BEHAVIOR_NOREPLY:
    ptr->flags.no_reply= data ? true : false;
    break;
  case MEMCACHED_BEHAVIOR_AUTO_EJECT_HOSTS:
    ptr->flags.auto_eject_hosts= data ? true : false;
    break;
  case MEMCACHED_BEHAVIOR_RANDOMIZE_REPLICA_READ:
      srandom((uint32_t) time(NULL));
      ptr->flags.randomize_replica_read= data ? true : false;
      break;
  default:
    /* Shouldn't get here */
    WATCHPOINT_ASSERT(flag);
    break;
  }

  return MEMCACHED_SUCCESS;
}

uint64_t memcached_behavior_get(memcached_st *ptr,
                                memcached_behavior_t flag)
{
  switch (flag)
  {
  case MEMCACHED_BEHAVIOR_NUMBER_OF_REPLICAS:
    return ptr->number_of_replicas;
  case MEMCACHED_BEHAVIOR_IO_MSG_WATERMARK:
    return ptr->io_msg_watermark;
  case MEMCACHED_BEHAVIOR_IO_BYTES_WATERMARK:
    return ptr->io_bytes_watermark;
  case MEMCACHED_BEHAVIOR_IO_KEY_PREFETCH:
    return ptr->io_key_prefetch;
  case MEMCACHED_BEHAVIOR_BINARY_PROTOCOL:
    return ptr->flags.binary_protocol;
  case MEMCACHED_BEHAVIOR_SUPPORT_CAS:
    return ptr->flags.support_cas;
  case MEMCACHED_BEHAVIOR_CACHE_LOOKUPS:
    return ptr->flags.use_cache_lookups;
    break;
  case MEMCACHED_BEHAVIOR_NO_BLOCK:
    return ptr->flags.no_block;
  case MEMCACHED_BEHAVIOR_BUFFER_REQUESTS:
    return ptr->flags.buffer_requests;
  case MEMCACHED_BEHAVIOR_USE_UDP:
    return ptr->flags.use_udp;
  case MEMCACHED_BEHAVIOR_TCP_NODELAY:
    return ptr->flags.tcp_nodelay;
  case MEMCACHED_BEHAVIOR_VERIFY_KEY:
    return ptr->flags.verify_key;
  case MEMCACHED_BEHAVIOR_KETAMA_WEIGHTED:
    return ptr->flags.ketama_weighted;
  case MEMCACHED_BEHAVIOR_DISTRIBUTION:
    return ptr->distribution;
  case MEMCACHED_BEHAVIOR_KETAMA:
    return (ptr->distribution == MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA) ? (uint64_t) 1 : 0;
  case MEMCACHED_BEHAVIOR_KETAMA_COMPAT_MODE:
    switch (ptr->distribution)
    {
    case MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA:
      return MEMCACHED_KETAMA_COMPAT_LIBMEMCACHED;
    case MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA_SPY:
      return MEMCACHED_KETAMA_COMPAT_SPY;
    case MEMCACHED_DISTRIBUTION_MODULA:
    case MEMCACHED_DISTRIBUTION_CONSISTENT:
    case MEMCACHED_DISTRIBUTION_RANDOM:
    default:
      return (uint64_t)-1;
    }
    /* NOTREACHED */
  case MEMCACHED_BEHAVIOR_HASH:
    return ptr->hash;
  case MEMCACHED_BEHAVIOR_KETAMA_HASH:
    return ptr->hash_continuum;
  case MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT:
    return ptr->server_failure_limit;
  case MEMCACHED_BEHAVIOR_SORT_HOSTS:
    return ptr->flags.use_sort_hosts;
  case MEMCACHED_BEHAVIOR_POLL_TIMEOUT:
    return (uint64_t)ptr->poll_timeout;
  case MEMCACHED_BEHAVIOR_CONNECT_TIMEOUT:
    return (uint64_t)ptr->connect_timeout;
  case MEMCACHED_BEHAVIOR_RETRY_TIMEOUT:
    return (uint64_t)ptr->retry_timeout;
  case MEMCACHED_BEHAVIOR_SND_TIMEOUT:
    return (uint64_t)ptr->snd_timeout;
  case MEMCACHED_BEHAVIOR_RCV_TIMEOUT:
    return (uint64_t)ptr->rcv_timeout;
  case MEMCACHED_BEHAVIOR_SOCKET_SEND_SIZE:
    {
      int sock_size;
      socklen_t sock_length= sizeof(int);

      /* REFACTOR */
      /* We just try the first host, and if it is down we return zero */
      if ((memcached_connect(&ptr->hosts[0])) != MEMCACHED_SUCCESS)
        return 0;

      if (getsockopt(ptr->hosts[0].fd, SOL_SOCKET,
                     SO_SNDBUF, &sock_size, &sock_length))
        return 0; /* Zero means error */

      return (uint64_t) sock_size;
    }
  case MEMCACHED_BEHAVIOR_SOCKET_RECV_SIZE:
    {
      int sock_size;
      socklen_t sock_length= sizeof(int);

      /* REFACTOR */
      /* We just try the first host, and if it is down we return zero */
      if ((memcached_connect(&ptr->hosts[0])) != MEMCACHED_SUCCESS)
        return 0;

      if (getsockopt(ptr->hosts[0].fd, SOL_SOCKET,
                     SO_RCVBUF, &sock_size, &sock_length))
        return 0; /* Zero means error */

      return (uint64_t) sock_size;
    }
  case MEMCACHED_BEHAVIOR_USER_DATA:
    return MEMCACHED_FAILURE;
  case MEMCACHED_BEHAVIOR_HASH_WITH_PREFIX_KEY:
    return ptr->flags.hash_with_prefix_key;
  case MEMCACHED_BEHAVIOR_NOREPLY:
    return ptr->flags.no_reply;
  case MEMCACHED_BEHAVIOR_AUTO_EJECT_HOSTS:
    return ptr->flags.auto_eject_hosts;
  case MEMCACHED_BEHAVIOR_RANDOMIZE_REPLICA_READ:
    return ptr->flags.randomize_replica_read;
  default:
    WATCHPOINT_ASSERT(flag);
    break;
  }

  WATCHPOINT_ASSERT(0); /* Programming mistake if it gets this far */
  return 0;
}
